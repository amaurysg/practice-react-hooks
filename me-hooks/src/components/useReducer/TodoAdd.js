import React from "react";
import useForm from "../../hooks/useForm";

const TodoAdd = ({ handleAddToDo }) => {
  //Here called CustomHooks useForm
  const [{ description }, handleInputChange, reset] = useForm({
    description: "",
  });
  // console.log("Description value",description)

  const handleSubmit = (e) => {
    e.preventDefault();

    //trim() delete spaces in strings
    if (description.trim().length <= 1) {
      //in this case do nothing
      return;
    }

    const newToDo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };

    //Eject function reset FROM useForm
    handleAddToDo(newToDo);
    reset();
  };

  return (
    <>
      <h4>Add Todo</h4>
      <hr />
      <form action="" onSubmit={handleSubmit}>
        <input
          className="form-control"
          type="text"
          name="description"
          value={description}
          onChange={handleInputChange}
          placeholder="Description to do..."
          autoComplete="off"
        />
        <button type="submit" className="btn btn-outline-primary mt-1 ">
          Add
        </button>
      </form>
    </>
  );
};

export default TodoAdd;
