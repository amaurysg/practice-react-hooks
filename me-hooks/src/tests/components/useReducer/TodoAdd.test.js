import { shallow } from "enzyme";
import TodoAdd from "../../../components/useReducer/TodoAdd";

describe("Test en componenten <TodoAdd/>", () => {
  const handleAddToDo = jest.fn();

  const wrapper = shallow(<TodoAdd handleAddToDo={handleAddToDo} />);

  test("Debe de mostrarse correctamente ", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("no debe llamar ADDTODO ", () => {
    //Validamos que no se dispare la funcion submit
    const formSubmit = wrapper.find("form").prop("onSubmit");
    // console.log(formSubmit);
    //Add preventDefault (debido a que en la funcion principal en el comp esta definido)
    formSubmit({ preventDefault() {} });
    //Esperamos que la funcion handleAddToDo haya sido llamada 0 veces
    expect(handleAddToDo).toHaveBeenCalledTimes(0);
  });
  test("Debe llamar la función handleAddToDo ", () => {
    const value = "React Jest Enzyme";
    wrapper.find("input").simulate("change", {
      target: {
        value,
        name: "description",
      },
    });
    const formSubmit = wrapper.find("form").prop("onSubmit");
    // console.log(formSubmit);
    //Add preventDefault (debido a que en la funcion principal en el comp esta definido)
    formSubmit({ preventDefault() {} });
    //Esperamos que la funcion handleAddToDo haya sido llamada 1 vez
    expect(handleAddToDo).toHaveBeenCalledTimes(1);
    expect(handleAddToDo).toHaveBeenCalledWith(expect.any(Object)); // { }
    //Se espera un objeto, pero id tiene que ser tipo Number
    expect(handleAddToDo).toHaveBeenCalledWith({
      id: expect.any(Number),
      desc: value,
      done: false,
    }); // { }

    //Test reset() value vació
    expect(wrapper.find("input").prop("value")).toBe("");
  });
});
