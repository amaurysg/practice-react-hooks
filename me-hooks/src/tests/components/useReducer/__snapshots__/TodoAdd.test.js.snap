// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test en componenten <TodoAdd/> Debe de mostrarse correctamente  1`] = `
<Fragment>
  <h4>
    Add Todo
  </h4>
  <hr />
  <form
    action=""
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Description to do..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 "
      type="submit"
    >
      Add
    </button>
  </form>
</Fragment>
`;
